### PR TITLE
8301374: NullPointerException in MemberEnter.checkReceiver

### PR DESCRIPTION
**Pull request by vicente-romero-oracle**
the following incorrect code is crashing javac:
```
public class X {
  interface F {
    void apply(E e);
  }
  enum E {
    ONE
  }

  F f = (E.ONE) -> {};   // E.ONE?
}
```
there are several issues the incorrect argument to this lambda expression. The proposed solution is to issue a syntax error for the missing parameter name. IMO, this error message would be correct here and this way the compiler fails fast instead of having to deal with an erroneous lambda later on to avoid a crash like the one we have now,

TIA

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301374](https://bugs.openjdk.org/browse/JDK-8301374): NullPointerException in MemberEnter.checkReceiver

### Reviewers
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12523/head:pull/12523` \
`$ git checkout pull/12523`

Update a local copy of the PR: \
`$ git checkout pull/12523` \
`$ git pull https://git.openjdk.org/jdk pull/12523/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12523`

View PR using the GUI difftool: \
`$ git pr show -t 12523`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12523.diff">https://git.openjdk.org/jdk/pull/12523.diff</a>

</details>

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java [hunk 1]
```diff
@@ -3723,6 +3723,10 @@ JCVariableDecl variableDeclaratorId(JCModifiers mods, JCExpression type, boolean
                 JCExpression pn = qualident(false);
                 if (pn.hasTag(Tag.IDENT) && ((JCIdent)pn).name != names._this) {
                     name = ((JCIdent)pn).name;
+                } else if (lambdaParameter) {
```
Comment (lahodaj) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java:3726]: I wonder if we could help the error recovery a little bit more, like e.g.:
```
                } else if (lambdaParameter && type == null) {
                    // we have a lambda parameter that is not an identifier this is a syntax error
                    type = pn;
                    name = names.empty;
                    reportSyntaxError(pos, Errors.Expected(IDENTIFIER));
                } else {
```

Which would keep the parsed thing as a type for the lambda parameter?

Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java]: yep I like this suggestion. I have updated the PR accordingly, thanks!

**Review by lahodaj — APPROVED** [review]
Looks good to me!

**Comment by vicente-romero-oracle** [comment]
@lahodaj thanks for the review

**Pull request closed without merging**